File: frontend/src/types/schemas.ts
import { z } from "zod";

// Zod schemas mirror the backend Pydantic models exactly. Every API response is
// validated at runtime against these, so a backend schema change surfaces as a
// type/validation error here instead of silently breaking at runtime.

export const strategyNameSchema = z.enum(["pomdp", "fixed", "self_rag", "adaptive"]);
export type StrategyName = z.infer<typeof strategyNameSchema>;

export const observationSchema = z.object({
  relevance_signal: z.number().min(0).max(1),
  evidence_coverage: z.number().min(0).max(1),
  confidence: z.number().min(0).max(1),
  source_response_time_ms: z.number().min(0),
  new_documents: z.number().int().min(0),
  duplicate_documents: z.number().int().min(0),
});

export const traceStepSchema = z.object({
  hop: z.number().int().min(0),
  action: z.string(),
  source: z.string().nullable(),
  prior_confidence: z.number().min(0).max(1),
  posterior_confidence: z.number().min(0).max(1),
  reward: z.number(),
  observation: observationSchema,
  query_used: z.string(),
  new_documents: z.number().int().min(0),
  duplicate_documents: z.number().int().min(0),
  injection_flagged: z.boolean(),
});
export type TraceStep = z.infer<typeof traceStepSchema>;

export const documentSchema = z.object({
  id: z.string(),
  title: z.string(),
  source: z.string(),
  score: z.number(),
  text: z.string(),
});

export const costSummarySchema = z.object({
  total_input_tokens: z.number().int().min(0),
  total_output_tokens: z.number().int().min(0),
  total_tokens: z.number().int().min(0),
  total_usd_cost: z.number().min(0),
  total_latency_ms: z.number().min(0),
  num_calls: z.number().int().min(0),
});

export const queryResponseSchema = z.object({
  query: z.string(),
  strategy: z.string(),
  answer: z.string(),
  final_confidence: z.number().min(0).max(1),
  hops: z.number().int().min(0),
  trace: z.array(traceStepSchema),
  cost: costSummarySchema,
  documents_used: z.array(documentSchema),
  terminal_reward: z.number(),
});
export type QueryResponse = z.infer<typeof queryResponseSchema>;

export const compareResponseSchema = z.object({
  query: z.string(),
  results: z.record(z.string(), queryResponseSchema),
});
export type CompareResponse = z.infer<typeof compareResponseSchema>;

export const strategyMetricsSchema = z.object({
  strategy: z.string(),
  accuracy: z.number(),
  avg_latency_ms: z.number(),
  avg_tokens: z.number(),
  avg_cost_usd: z.number(),
  avg_hops: z.number(),
  efficiency: z.number(),
});
export type StrategyMetrics = z.infer<typeof strategyMetricsSchema>;

export const batchEvalResponseSchema = z.object({
  count: z.number().int().min(0),
  metrics: z.array(strategyMetricsSchema),
});
export type BatchEvalResponse = z.infer<typeof batchEvalResponseSchema>;

export const tokenResponseSchema = z.object({
  access_token: z.string(),
  token_type: z.string(),
  expires_in: z.number().int(),
});

export const userInfoSchema = z.object({
  username: z.string(),
  issuer: z.string(),
});
export type UserInfo = z.infer<typeof userInfoSchema>;

export const healthResponseSchema = z.object({
  status: z.literal("ok"),
  mode: z.enum(["local", "cloud"]),
  strategies: z.array(z.string()),
  corpus_size: z.number().int().min(0),
});

export const errorResponseSchema = z.object({
  error_code: z.string(),
  message: z.string(),
  details: z.record(z.string(), z.unknown()).nullable().optional(),
});

File: backend/app/schemas.py
"""Pydantic v2 request/response schemas with exhaustive validation.

Validation is expressed with explicit ``Field`` constraints (not just type
hints): bounded numerics, length-limited and control-char-free text, ``Literal``
strategy values, and a capped batch size to prevent cost-abuse. Every API error
is returned in a consistent structured shape (see ``ErrorResponse``).
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field, field_validator

from app.security.sanitize import strip_control_chars

StrategyName = Literal["pomdp", "fixed", "self_rag", "adaptive"]

MAX_QUERY_LEN = 2000


def _clean_query(value: str) -> str:
    cleaned = strip_control_chars(value).strip()
    if not cleaned:
        raise ValueError("query must not be empty after trimming")
    return cleaned


# --- observation / trace ----------------------------------------------------
class ObservationSchema(BaseModel):
    relevance_signal: float = Field(ge=0.0, le=1.0)
    evidence_coverage: float = Field(ge=0.0, le=1.0)
    confidence: float = Field(ge=0.0, le=1.0)
    source_response_time_ms: float = Field(ge=0.0)
    new_documents: int = Field(ge=0)
    duplicate_documents: int = Field(ge=0)


class TraceStepSchema(BaseModel):
    hop: int = Field(ge=0)
    action: str
    source: str | None = None
    prior_confidence: float = Field(ge=0.0, le=1.0)
    posterior_confidence: float = Field(ge=0.0, le=1.0)
    reward: float
    observation: ObservationSchema
    query_used: str
    new_documents: int = Field(ge=0)
    duplicate_documents: int = Field(ge=0)
    injection_flagged: bool = False


class DocumentSchema(BaseModel):
    id: str
    title: str = ""
    source: str
    score: float
    text: str


class CostSummarySchema(BaseModel):
    total_input_tokens: int = Field(ge=0)
    total_output_tokens: int = Field(ge=0)
    total_tokens: int = Field(ge=0)
    total_usd_cost: float = Field(ge=0.0)
    total_latency_ms: float = Field(ge=0.0)
    num_calls: int = Field(ge=0)


# --- query ------------------------------------------------------------------
class QueryRequest(BaseModel):
    query: str = Field(min_length=1, max_length=MAX_QUERY_LEN)
    strategy: StrategyName = "pomdp"

    @field_validator("query")
    @classmethod
    def _validate_query(cls, value: str) -> str:
        return _clean_query(value)


class QueryResponse(BaseModel):
    query: str
    strategy: str
    answer: str
    final_confidence: float = Field(ge=0.0, le=1.0)
    hops: int = Field(ge=0)
    trace: list[TraceStepSchema]
    cost: CostSummarySchema
    documents_used: list[DocumentSchema]
    terminal_reward: float


class CompareResponse(BaseModel):
    query: str
    results: dict[str, QueryResponse]


# --- batch evaluation -------------------------------------------------------
class EvalItem(BaseModel):
    query: str = Field(min_length=1, max_length=MAX_QUERY_LEN)
    reference: str = Field(min_length=1, max_length=MAX_QUERY_LEN)

    @field_validator("query", "reference")
    @classmethod
    def _clean(cls, value: str) -> str:
        return _clean_query(value)


class BatchEvalRequest(BaseModel):
    # Cap batch size to prevent cost-abuse from a single call.
    items: list[EvalItem] = Field(min_length=1, max_length=50)
    strategies: list[StrategyName] | None = None


class StrategyMetricsSchema(BaseModel):
    strategy: str
    accuracy: float
    avg_latency_ms: float
    avg_tokens: float
    avg_cost_usd: float
    avg_hops: float
    efficiency: float


class BatchEvalResponse(BaseModel):
    count: int = Field(ge=0)
    metrics: list[StrategyMetricsSchema]


# --- auth -------------------------------------------------------------------
class LoginRequest(BaseModel):
    username: str = Field(min_length=1, max_length=128)
    password: str = Field(min_length=1, max_length=256)


class TokenResponse(BaseModel):
    access_token: str
    token_type: str = "bearer"
    expires_in: int = Field(ge=1)


class UserInfo(BaseModel):
    username: str
    issuer: str


# --- health / errors --------------------------------------------------------
class HealthResponse(BaseModel):
    status: Literal["ok"] = "ok"
    mode: Literal["local", "cloud"]
    strategies: list[str]
    corpus_size: int = Field(ge=0)


class ErrorResponse(BaseModel):
    error_code: str
    message: str
    details: dict | None = None
